### ISSUE TITLE
Bias in model

### ISSUE THREAD

**Issue by Haroenv** [infinitered/nsfwjs#16]
There seems to be a bias of when an image contains a woman with _any_ skin showing, the model will mark those as porn, even when they clearly aren't.

There's some examples in [this twitter post](https://twitter.com/_Nec/status/1098140532751118338), but I could reproduce it with different images too.

**Comment by catdad** [comment]
I am finding very similar issues as well:

![image](https://user-images.githubusercontent.com/2205537/53109390-c6080100-3506-11e9-8680-3e1ada6dbb46.png)

**Comment by GantMan** [comment]
Agreed.  I have a plan to help fix this.  One thing to note is that faces near microphones might exacerbate this bias.  The original training data was provided by https://github.com/alexkimxyz/nsfw_data_scrapper

The input images seem to slightly off due to the subreddits that fueled the dataset. Hopefully, I can counterbalance this and contribute back to that original repo for all.

**Comment by Bengejd** [comment]
I tried going the opposite approach and seeing how it classified women who were predominantly covered up, I used two different types of women, two with women in traditional Muslim Garb, and one with traditional Quaker garb. I then tried out an image that (by my analysis) would be considered more "sexy", and the results were woefully incorrect.

![collage](https://user-images.githubusercontent.com/11723093/53112657-c8218e00-350d-11e9-9f84-04f1f15effa0.png)

As you can see, the predominantly covered group, are all flagged as porn with a high certainty, despite being anything but. Though interestingly, they are also all flagged with a high "Neutral" rating as well.  

VS this woman in a revealing, but techincally not "porn quality" bikini.

<img width="314" alt="screen shot 2019-02-20 at 12 39 07 pm" src="https://user-images.githubusercontent.com/11723093/53112689-da033100-350d-11e9-89b7-5c919e99cf21.png">

This would be a good lesson for the ML Model, as you could use this result to test false positives, where if the "Porn" & "Sexy" categories are the two highest, it is most likely porn. Obviously, this would still allow for some false positives labeled as "porn" to fall through the cracks, but it's a step in the right direction. Similarly, if the two highest categories are "Porn" and "Neutral", then there is a good chance that it is also a false positive as well.

**Comment by GantMan** [comment]
Great examples so far!  This perspective is very valuable.  I filed an issue on the data-repo.

#### Side note:
I've gotten lots of requests to share the dataset (for a bunch of reasons, not all for science).   But with some public curation of the training data, the model can significantly improve.  Adding a few hundred photos of false positives into the test/train will help.  

As a call to action, I would love to crowsource-fix this problem.  Feel free to let me know you have a zip of false-positive images to add, or if you'd be interested in curation.

**Comment by alexkimxyz** [comment]
> ... The original training data was provided by https://github.com/alexkimxyz/nsfw_data_scrapper
@GantMan it'd be nice if you credited the original repo https://github.com/alexkimxyz/nsfw_data_scraper in both your medium blog post and this repo.

**Comment by GantMan** [comment]
You read my mind @alexkimxyz - you've been added to both! Sorry for the delay

**Comment by MathiasaurusRex** [comment]
Bearded man covered in snow:

<img width="544" alt="screen shot 2019-02-20 at 8 26 04 pm" src="https://user-images.githubusercontent.com/3649460/53138992-c677ba80-354d-11e9-8ef7-367460c99f48.png">

**Comment by Jrexexe** [comment]
My fingers 
![微信图片_20190222112357.png](https://i.loli.net/2019/02/22/5c6f6b91cfcd2.png)

**Comment by GantMan** [comment]
Hah, is there a subreddit of hands?  If so we can add it.

**Comment by Haroenv** [comment]
Some more examples here: https://twitter.com/JustTenDads/status/1098502194196697088 (Jeffrey goldblum)

**Comment by GantMan** [comment]
RE: Jeff Goldblum

![image](https://user-images.githubusercontent.com/997157/53252118-d5608900-3683-11e9-9d82-ad1f2f5234c4.png)

🤣

**Comment by GantMan** [comment]
## MODEL BIAS UPDATE:
**I'm working on a newly trained model.  This takes DAYS on my home computer.  This will take time**

I appreciate everyone's feedback!  You'll be happy to know I'm grabbing some updated training photos to hopefully help fix this bias... except for Jeff Goldblum, that stays.   I'm going to lock this conversation for two reasons, which I hope are fair and clear.

1.  I'm actively working on an update which will have new biases, so noted biases on the existing model won't help.  A new ticket can be created when the new model is in place.
2. This model is trained on GIGS of data, so single examples actually don't help as much as providing an excellent subreddit to the data scraper repo.   The workflow is like so:

### [NSFW Data Scraper](https://github.com/alexkimxyz/nsfw_data_scrapper) ➡️ [NSFW Model](https://github.com/gantman/nsfw_model)  ➡️  NSFW JS 🥳

So if you find category bias, please contribute back to the data scraper, it's easy!  Those contributions will make it to the model, which will make it to NSFW JS.

**Comment by GantMan** [comment]
I'm happy to announce after a lot of tweaking, adjusting, and hours of additional training.  I've increased the test set to reflect a broader array of images, AND increased accuracy to 93% on that larger dataset.   

## While model bias is unavoidable, and accuracy is below a human's potential
I'm very happy to continue working on improving NSFW JS

Thank you all for your feedback, and I hope you find the following results pleasing.

![image](https://user-images.githubusercontent.com/997157/53823053-97424f80-3f36-11e9-9f69-7659991dae0d.png)

### I consider this more accurate than the original classification
![image](https://user-images.githubusercontent.com/997157/53823080-a88b5c00-3f36-11e9-8901-58d38831c178.png)

![image](https://user-images.githubusercontent.com/997157/53823108-b5a84b00-3f36-11e9-96d3-e517696d5d35.png)

![image](https://user-images.githubusercontent.com/997157/53823144-c658c100-3f36-11e9-8d4f-4bcb30b49e4a.png)

![image](https://user-images.githubusercontent.com/997157/53823177-d53f7380-3f36-11e9-858c-a4f09b8822b0.png)

![image](https://user-images.githubusercontent.com/997157/53823197-e4262600-3f36-11e9-8fe6-d0134629f5bb.png)

![image](https://user-images.githubusercontent.com/997157/53823402-51d25200-3f37-11e9-8101-2c074bf1d0b2.png)


### And unfortunately, 1 false positive.   Which is of-course to be expected.  93% accuracy is not nearly as good as a human.
![image](https://user-images.githubusercontent.com/997157/53823494-83e3b400-3f37-11e9-9412-136af9d43a3e.png)

If someone can suggest a large set of images of people singing into microphones, I'd love to add that to the training data.  Please supply zip files with hundreds of examples in appropriate folders, if you'd like to help!

I'd like to thank everyone who helped in the spirit of making something creative and useful.  Please keep in mind this is not my full-time job.  I build things like this as a passion to help the community.

This model is about to be released.  If you have old results, please clear your cache and make sure you get the latest model.

**Comment by GantMan** [comment]
## BONUS:   Jeff Goldblum is fixed... also I'm sad about this 😿

![image](https://user-images.githubusercontent.com/997157/53824387-73343d80-3f39-11e9-9235-1261e69b75d2.png)


![image](https://user-images.githubusercontent.com/997157/53824362-64e62180-3f39-11e9-9be0-411cca1122de.png)

![image](https://user-images.githubusercontent.com/997157/53824415-83e4b380-3f39-11e9-84f9-e400d2d74e38.png)

![image](https://user-images.githubusercontent.com/997157/53824446-9363fc80-3f39-11e9-9774-8eb3df3149c6.png)
